hpat/datatypes/hpat_pandas_series_functions.py

```diff
@@ -165,6 +200,174 @@ def hpat_pandas_series_shape_impl(self):
+    Pandas Series method :meth:`pandas.Series.nsmallest` implementation.
+
+    .. only:: developer
+       Test: python -m hpat.runtests -k hpat.tests.test_series.TestSeries.test_series_nsmallest*
+
+    Parameters
+    ----------
+    self: :obj:`pandas.Series`
+        input series
+    n: :obj:`int`, default 5
+        Return this many ascending sorted values.
+    keep: :obj:`str`, default 'first'
+        When there are duplicate values that cannot all fit in a Series of n elements:
+        first : return the first n occurrences in order of appearance.
+        last : return the last n occurrences in reverse order of appearance.
+        all : keep all occurrences. This can result in a Series of size larger than n.
+        *unsupported*
+
+    Returns
+    -------
+    :obj:`series`
+         returns :obj:`series`
+    """
+
+    _func_name = 'Method nsmallest().'
+
+    ty_checker = TypingChecker(_func_name)
+    ty_checker.check_series(self)
+    ty_checker.check_integer(n, 'n')
+    ty_checker.check_string(keep, 'keep')
+
+    dtype = self.data.dtype
+    if isinstance(self.index, types.NoneType):
+        inner_list_type = types.ListType(types.intp)
+        def hpat_pandas_series_nsmallest_impl(self, n=5, keep='first'):
+            if keep != 'first':
+                raise ValueError("Method nsmallest(). Unsupported parameter. "
+                                 "Given 'keep' != 'first'")
+
+            nlargest = hpat.hiframes.api.nlargest(self._data, n, False,
+                                                  hpat.hiframes.series_kernels.lt_f)
+            indices = Dict.empty(dtype, inner_list_type)
+            for item in set(nlargest):
+                indices[item] = List.empty_list(types.intp)
+
+            for idx, item in enumerate(self._data):
+                if item in indices:
+                    indices[item].append(idx)
+
+            local_index = [idx for item in sorted(indices) for idx in indices[item]]
+            local_index = local_index[:len(nlargest)]
+
+            return pandas.Series(nlargest, local_index, name=self._name)
+
+        return hpat_pandas_series_nsmallest_impl
+
+    index_type = self.index.dtype
+    inner_list_type = types.ListType(index_type)
+    def hpat_pandas_series_nsmallest_impl(self, n=5, keep='first'):
+        if keep != 'first':
+            raise ValueError("Method nsmallest(). Unsupported parameter. "
+                             "Given 'keep' != 'first'")
+
+        nlargest = hpat.hiframes.api.nlargest(self._data, n, False,
+                                              hpat.hiframes.series_kernels.lt_f)
+        indices = Dict.empty(dtype, inner_list_type)
+        for item in set(nlargest):
+            indices[item] = List.empty_list(index_type)
+
+        for idx, item in enumerate(self._data):
+            if item in indices:
+                indices[item].append(self._index[idx])
+
+        local_index = [idx for item in sorted(indices) for idx in indices[item]]
```

Comment:
In theory we can have situation when all our Big Array is filled with 5 numbers and with this loop comprehension we will have to iterate over it once again, i.e. what we lack is the ability to leave early after we found exactly n first indices. What do you think if we use explicit loop here?
```suggestion
            local_index = []
            for item in nsmallest:
                for index in indices[item]:
                    local_index.append(index)
                    if len(local_index) == n:
                        break
                else:
                    continue
                break
```

Reply:
Done.